VaadinPlugin does not work in hotswapagent 1.4.1 on Wildfly 19 on Windows 10

I just installed the latest dcevm for java 11 and it is using hotswapagent 1.4.1 This version includes an updated VaadinPlugin, which I was hoping would be an improvement. However on Wildfly 19 (which I am using) it does not work with the following exception:

> 11:30:12,897 ERROR [io.undertow.request:] (default task-6) UT005023: Exception handling request to /webportfolio/push: javax.servlet.ServletException: UT010013: Could not instantiate com.vaadin.cdi.CdiServletDeployer
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.core.ManagedServlet$DefaultInstanceStrategy.start(ManagedServlet.java:302)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.core.ManagedServlet.forceInit(ManagedServlet.java:212)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletChain.forceInit(ServletChain.java:130)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:63)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.core@2.0.30.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.core@2.0.30.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.core@2.0.30.Final//io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.core@2.0.30.Final//io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.core@2.0.30.Final//io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.core@2.0.30.Final//io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.core@2.0.30.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.core@2.0.30.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.security.jaspi.JASPICSecureResponseHandler.handleRequest(JASPICSecureResponseHandler.java:48)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at io.undertow.core@2.0.30.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:269)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:78)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:133)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:130)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1541)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1541)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1541)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1541)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:249)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:78)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:99)
	at io.undertow.core@2.0.30.Final//io.undertow.server.Connectors.executeRootHandler(Connectors.java:376)
	at io.undertow.core@2.0.30.Final//io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads@2.3.3.Final//org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads@2.3.3.Final//org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982)
	at org.jboss.threads@2.3.3.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads@2.3.3.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: javax.enterprise.inject.CreationException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.security.NewInstanceAction.run(NewInstanceAction.java:33)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.Exceptions.rethrowException(Exceptions.java:40)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.Exceptions.rethrowException(Exceptions.java:50)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.Exceptions.rethrowException(Exceptions.java:90)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.ConstructorInjectionPoint.newInstance(ConstructorInjectionPoint.java:127)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.ConstructorInjectionPoint.invokeAroundConstructCallbacks(ConstructorInjectionPoint.java:92)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.ConstructorInjectionPoint.newInstance(ConstructorInjectionPoint.java:78)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.producer.AbstractInstantiator.newInstance(AbstractInstantiator.java:28)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.producer.BasicInjectionTarget.produce(BasicInjectionTarget.java:112)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.producer.BeanInjectionTarget.produce(BeanInjectionTarget.java:186)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.bean.ManagedBean.create(ManagedBean.java:160)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.contexts.unbound.DependentContextImpl.get(DependentContextImpl.java:64)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:694)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:717)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.util.ForwardingBeanManager.getReference(ForwardingBeanManager.java:64)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.bean.builtin.BeanManagerProxy.getReference(BeanManagerProxy.java:87)
	at org.jboss.as.weld@19.0.0.Final//org.jboss.as.weld.deployment.WeldClassIntrospector$1.getReference(WeldClassIntrospector.java:79)
	at org.wildfly.extension.undertow@19.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$ManagedReferenceInstanceFactory.createInstance(UndertowDeploymentInfoService.java:1341)
	at io.undertow.servlet@2.0.30.Final//io.undertow.servlet.core.ManagedServlet$DefaultInstanceStrategy.start(ManagedServlet.java:300)
	... 43 more
Caused by: java.lang.Error: java.lang.reflect.InvocationTargetException
	at deployment.webportfolio.war//com.vaadin.flow.server.VaadinServlet.<init>(VaadinServlet.java:51)
	at deployment.webportfolio.war//com.vaadin.cdi.CdiVaadinServlet.<init>(CdiVaadinServlet.java:46)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.jboss.weld.core@3.1.3.Final//org.jboss.weld.injection.ConstructorInjectionPoint.newInstance(ConstructorInjectionPoint.java:119)
	... 59 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	... 66 more
Caused by: java.lang.NoClassDefFoundError: com/vaadin/flow/server/VaadinService
	at java.base/java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredConstructors(Class.java:3137)
	at java.base/java.lang.Class.getConstructor0(Class.java:3342)
	at java.base/java.lang.Class.getConstructor(Class.java:2151)
	at org.hotswap.agent.plugin.vaadin.VaadinPlugin.registerServlet(VaadinPlugin.java:85)
	... 70 more
Caused by: java.lang.ClassNotFoundException: com.vaadin.flow.server.VaadinService
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 75 more

It seems that the  `org.hotswap.agent.plugin.vaadin.VaadinPlugin.registerServlet()` method has a problem. the appClassLoader.loadClass statement on line 83 returns a class instance with the ClassLoaders$AppClassLoader class loader (internal classloader in the JDK). Instead it should have used the appClassLoader that was used to load it (in this case the war-specific moduleclassloader from Wildfly). 

The following change fixes my problem (notice I had to change to Class.forName AND change to explicit string):
```
83,84c83,84
<             Class<?> vaadinIntegrationClass = appClassLoader.loadClass(
<                     VaadinIntegration.class.getName());
---
>             Class<?> vaadinIntegrationClass = Class.forName( "org.hotswap.agent.plugin.vaadin.VaadinIntegration",
>                     true, appClassLoader);
```

--- a/plugin/hotswap-agent-vaadin-plugin/src/main/java/org/hotswap/agent/plugin/vaadin/VaadinIntegration.java
+++ b/plugin/hotswap-agent-vaadin-plugin/src/main/java/org/hotswap/agent/plugin/vaadin/VaadinIntegration.java
@@ -13,6 +13,7 @@ import org.hotswap.agent.logging.AgentLogger;
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.router.Route;
 import com.vaadin.flow.router.RouteConfiguration;
+import com.vaadin.flow.router.internal.RouteUtil;
 import com.vaadin.flow.server.RouteRegistry;
 import com.vaadin.flow.server.VaadinContext;
 import com.vaadin.flow.server.VaadinService;
@@ -149,9 +150,10 @@ public class VaadinIntegration {
                     .filter(clazz -> clazz.isAnnotationPresent(Route.class))
                     .forEach(clazz -> {
                         Class<? extends Component> componentClass = (Class<? extends Component>) clazz;
+
+                        final Route route = componentClass.getAnnotation(Route.class);
                         LOGGER.info(
-                                "Updating route '{}' to {}", componentClass
-                                        .getAnnotation(Route.class).value(),
+                                "Updating route '{}' to {}", RouteUtil.resolve(componentClass, route),
                                 clazz);
                         routeConf.removeRoute(componentClass);
                         routeConf.setAnnotatedRoute(componentClass);

--- a/plugin/hotswap-agent-vaadin-plugin/src/main/java/org/hotswap/agent/plugin/vaadin/VaadinPlugin.java
+++ b/plugin/hotswap-agent-vaadin-plugin/src/main/java/org/hotswap/agent/plugin/vaadin/VaadinPlugin.java
@@ -27,11 +27,11 @@ import org.hotswap.agent.util.PluginManagerInvoker;
  *
  * @author Artur Signell
  * @author Matti Tahvonen
- * @autho Johannes Eriksson
+ * @author Johannes Eriksson
  */
 @Plugin(name = "Vaadin",
         description = "Vaadin Platform support",
-        testedVersions = {"14.1.20", "14.2.0.alpha7", "15.0.2"},
+        testedVersions = {"14.1.20", "14.2.0.alpha7", "14.3.0.beta2", "15.0.2"},
         expectedVersions = {"14 - 16"})
 public class VaadinPlugin {
 
@@ -80,12 +80,10 @@ public class VaadinPlugin {
 
     public void registerServlet(Object vaadinServlet) {
         try {
-            Class<?> vaadinIntegrationClass = appClassLoader.loadClass(
-                    VaadinIntegration.class.getName());
+            Class<?> vaadinIntegrationClass = resolveClass("org.hotswap.agent.plugin.vaadin.VaadinIntegration");
             Object vaadinIntegration = vaadinIntegrationClass.getConstructor()
                     .newInstance();
-            Class<?> vaadinServletClass = Class.forName("com.vaadin.flow.server.VaadinServlet",
-                    true, appClassLoader);
+            Class<?> vaadinServletClass = resolveClass("com.vaadin.flow.server.VaadinServlet");
             Method m = vaadinIntegrationClass.getDeclaredMethod("servletInitialized",
                     vaadinServletClass);
             m.invoke(vaadinIntegration, vaadinServlet);

